File: PlaineDFT/example/debug_nonloc/test_nonloc.py
import os
os.environ['OMP_NUM_THREADS'] = '2'
import numpy as np
from numpy.random import seed, rand
from plainedft import Atoms
from plainedft import SCF
from plainedft.scf import getE, orth#, orth2
from plainedft.plot import *
from plainedft.tools import *
from plainedft.gth_nonloc import calc_Enl, calc_Vnl


def main(name):
    if name == 'Ne':
        atom = 'Ne'
        lattice = 16
        X = np.array([[8.000000003522448, 8.000000003522448, 8.000000003522448]])
        Z = 8
        Ns = 4
        S = 75 * np.array([1, 1, 1])
        f = 2
        ecut = 25
        verbose = 5
        pot = 'gth'

    if name == 'LiH':
        atom = ['Li', 'H']
        lattice = 16
        X = np.array([[0,0,0], [8.000000003522448, 8.000000003522448, 8.000000003522448]])
        Z = [1, 1]
        Ns = 1
        S = 75 * np.array([1, 1, 1])
        f = 2
        ecut = 25
        verbose = 5
        pot = 'gth'

    # Create atoms object, with pseudopotential parameters inside
    atoms = Atoms(atom=atom, a=lattice, X=X, Z=Z, Ns=Ns, S=S, f=f, ecut=ecut, verbose=verbose, pot=pot)

    # Create a basis set with a set seed
    seed(1234)
    W = rand(len(atoms.active[0]), atoms.Ns)
    # W = orth(atoms, W)

    E = calc_Enl(atoms, W)
    V = calc_Vnl(atoms, W)

    betaNL_psi = np.dot(W.T.conj(), atoms.betaNL).conj()

    return atoms, E, V, W, betaNL_psi

if __name__ == '__main__':
    from timeit import default_timer
    start = default_timer()
    print(f'Enl: {main("LiH")[1]}')
    end = default_timer()
    print(f'Time: {end-start}')

File: PlaineDFT/example/debug_nonloc/test_nonloc.jl
using LinearAlgebra
using Random
using Printf
using PWDFT
using PyCall

const DIR_PWDFT = joinpath(dirname(pathof(PWDFT)),"..")
const DIR_PSP = joinpath(DIR_PWDFT, "pseudopotentials", "pade_gth")
#const DIR_STRUCTURES = joinpath(DIR_PWDFT, "structures")

# Needed to build a wave function, bc I am too stupid to get the correct shape
include("/home/wanja/PWDFT.jl/src/gen_wavefunc.jl")

function main(name)
    Random.seed!(1234)

    if name == "Ne"
        # Atoms
        atoms = Atoms( xyz_file=joinpath("", "Ne.xyz"),
                       LatVecs = gen_lattice_sc(16.0) )

        # Initialize Hamiltonian
        pspfiles = [joinpath(DIR_PSP, "Ne-q8.gth")]
        ecutwfc = 25.0
        Ham = Hamiltonian( atoms, pspfiles, ecutwfc )

        # Create a basis set with a set seed, same as in python
        py"""
        from numpy.random import rand, seed
        seed(1234)
        # Has to have the same form as the python version (Number of G-vectors, States)
        xs = rand(24405, 4)
        """
        xs = PyArray(py"xs"o)
    end

    if name == "LiH"
        # Atoms
        atoms = Atoms( xyz_file=joinpath("", "LiH.xyz"),
                       LatVecs = gen_lattice_sc(16.0) )

        # Initialize Hamiltonian
        pspfiles = [joinpath(DIR_PSP, "Li-q1.gth"), joinpath(DIR_PSP, "H-q1.gth")]
        ecutwfc = 25.0
        Ham = Hamiltonian( atoms, pspfiles, ecutwfc )

        # Create a basis set with a set seed, same as in python
        py"""
        from numpy.random import rand, seed
        seed(1234)
        # Has to have the same form as the python version (Number of G-vectors, States)
        xs = rand(24405, 1)
        """
        xs = PyArray(py"xs"o)
    end

    # display(Ham)
    # Create wfc (just to get the correct shape)
    psiks = rand_wfc( Ham )
    psiks[1] = xs
    # psiks[1] = ortho_sqrt(psiks[1])

    E = calc_energies( Ham, psiks )
    V = op_V_Ps_nloc( Ham, psiks )

    betaNL_psi = calc_betaNL_psi( 1, Ham.pspotNL.betaNL, psiks[1] )

    return Ham, E.Ps_nloc, V, psiks[1], betaNL_psi
end

main("LiH")
